pmd-java/etc/grammar/Java.jjt

```diff
@@ -1669,18 +1710,25 @@ void InstanceOfExpression() #void:
   RelationalExpression() [
     LOOKAHEAD(1)
     ("instanceof"
-        AnnotatedRefType() [ BindingVarId() #TypeTestPattern(2) ]
+        (
+            AnnotatedRefType() [ VariableDeclaratorId() #TypePattern(2) ]
+            |
+            ( LocalVarModifierList() ReferenceType() VariableDeclaratorId() ) #TypePattern(3)
+        )
         {
             jjtThis.setOp(BinaryOp.INSTANCEOF);
             AbstractJavaNode top = jjtree.popNode();
             if (top instanceof ASTPattern) {
+               if (top.getNumChildren() == 2) {
+                   insertEmptyModifierListWithAnnotations(top, (AbstractJavaNode) top.getChild(0));
+               }
                top = new ASTPatternExpression((ASTPattern) top);
             } else {
                top = new ASTTypeExpression((ASTType) top);
             }
             jjtree.pushNode(top);
         }
-        {}
+        {} // seems to be important?
```

Comment:
This is because JJTree has special handling for a java block that is the very last thing in a node scope. Here this block is the last one in the scope of `#InfixExpression` below.

The difference is that in that last block, the `jjtClose` routine, and other closing code has been executed, including pushing the node on the stack, and popping its children. This allows you to write `{ return jjtThis; }` as the last thing in a production (while being assured that the node is fully built before the return statement).

This also means that if we want to do some transformation on the children before the node is closed, we cannot do it in the last block. So we insert this dummy empty block to be the last one, and our logic is in the next to last block so executes before the node is closed.

Reply:
Ah, thanks for the explanation! I only noticed, that the tree isn't built correctly, if I remove that :smile: